zquery/shared/src/main/scala/zio/zquery/BlockedRequests.scala

```diff
@@ -0,0 +1,210 @@
+   * request type of each data source.
+   */
+  final def mapDataSources[R1](f: DataSourceFunction[R, R1]): BlockedRequests[R1] =
+    self match {
+      case Empty            => empty
+      case Parallel(l, r)   => parallel(l.mapDataSources(f), r.mapDataSources(f))
+      case Sequential(l, r) => sequential(l.mapDataSources(f), r.mapDataSources(f))
+      case Single(ds, br)   => single(f(ds), br)
+    }
+
+  /**
+   * Executes all requests, submitting batched requests to each data source in
+   * parallel.
+   */
+  val run: ZIO[R, Nothing, Unit] = {
+    val flattened = BlockedRequests.flatten(self)
+    val pipelined = BlockedRequests.pipeline(flattened)
+    ZIO.foreach_(pipelined) { requests =>
+      val grouped = BlockedRequests.groupByDataSource(requests)
+      ZIO.foreachPar_(grouped) {
+        case (dataSource, requests) =>
+          val requestsToRun = requests.map(_.map(_.request))
+          for {
+            completedRequests <- dataSource.runAll(requestsToRun)
+            _ <- ZIO.foreach_(requests) { blockedRequest =>
+                  ZIO.foreach(blockedRequest) { br =>
+                    br.result.set(completedRequests.lookup(br.request))
+                  }
+                }
+          } yield ()
+      }
+    }
+  }
+}
+
+private[zquery] object BlockedRequests {
+
+  /**
+   * The empty set of blocked requests.
+   */
+  val empty: BlockedRequests[Any] =
+    Empty
+
+  /**
+   * Constructs a set of blocked requests from two sets of blocked requests
+   * that can be performed in parallel.
+   */
+  def parallel[R](left: BlockedRequests[R], right: BlockedRequests[R]): BlockedRequests[R] =
+    Parallel(left, right)
+
+  /**
+   * Constructs a set of blocked requests from two sets of blocked requests
+   * that must be performed sequentially.
+   */
+  def sequential[R](left: BlockedRequests[R], right: BlockedRequests[R]): BlockedRequests[R] =
+    Sequential(left, right)
+
+  /**
+   * Constructs a set of blocked requests from the specified blocked request
+   * data data source.
+   */
+  def single[R, K](dataSource: DataSource[R, K], blockedRequest: BlockedRequest[K]): BlockedRequests[R] =
+    Single(dataSource, blockedRequest)
+
+  case object Empty extends BlockedRequests[Any]
+
+  final case class Single[-R, A](dataSource: DataSource[R, A], blockedRequest: BlockedRequest[A])
+      extends BlockedRequests[R]
+
+  final case class Parallel[-R](left: BlockedRequests[R], right: BlockedRequests[R]) extends BlockedRequests[R]
+
+  final case class Sequential[-R](left: BlockedRequests[R], right: BlockedRequests[R]) extends BlockedRequests[R]
+
+  /**
+   * Flattens a collection of blocked requests to a sequence of sets of
+   * blocked requests, where each set represents requests that can be performed
+   * in parallel and sequential sets represent requests that must be performed
+   * sequentially.
+   */
+  private def flatten[R](blockedRequests: BlockedRequests[R]): List[Set[Single[R, _]]] = {
```

Comment:
I suggest start moving to `Chunk` and away from `List` for most purposes...

Reply:
Good idea.